### PR TITLE
a send task or message throw events can reference publish message

### LINKED ISSUE #12878: Add message definition extension elements
**Is your feature request related to a problem? Please describe.**
In the BPMN model API, a `send task` or `message throw event` should be able to reference a message definition.

* a `send task` or `message throw event` can have an extension element `messageDefinition` with
  * an attribute `messageName` that references the name of the message
    * has either a static value or an expression
    * must not be empty
  * an attribute `correlationKey` that references the correlation key of the message
    * has either a static value or an expression
    * must not be empty
  * an attribute `messageId` that references the id of the message
    * has either a static value or an expression
  * an attribute `timeToLive` that references the time to live of the message
    * has either a static value or an expression
*  a `send task` or `message throw event` must have either an extension element `messageDefinition` or `taskDefinition` but not both

Sample XML with static attributes:

```
<bpmn:sendTask id="send_message">
  <bpmn:extensionElements>
    <zeebe:messageDefinition 
        messageName="foo" 
        correlationKey="abc" 
        messageId="abc" 
        timeToLive="PT10S"/>
  </bpmn:extensionElements>  
</bpmn:sendTask> 


<bpmn:intermediateThrowEvent id="Event_1nqoocy">
  <bpmn:extensionElements>
    <zeebe:messageDefinition 
        messageName="foo" 
        correlationKey="abc" 
        messageId="abc" 
        timeToLive="PT10S"/>
  </bpmn:extensionElements>
  <bpmn:messageEventDefinition id="MessageEventDefinition_1wizp0b" />
</bpmn:intermediateThrowEvent>
```

Sample XML with expression attributes:

```
<bpmn:sendTask id="send_message">
  <bpmn:extensionElements>
    <zeebe:messageDefinition 
        messageName="=messageName" 
        correlationKey="=correlationKey" 
        messageId="=messageId" 
        timeToLive="=timeToLive"/>
  </bpmn:extensionElements>  
</bpmn:sendTask> 

<bpmn:intermediateThrowEvent id="Event_1nqoocy">
  <bpmn:extensionElements>
    <zeebe:messageDefinition 
        messageName="=messageName" 
        correlationKey="=correlationKey" 
        messageId="=messageId" 
        timeToLive="=timeToLive"/>
  </bpmn:extensionElements>
  <bpmn:messageEventDefinition id="MessageEventDefinition_1wizp0b" />
</bpmn:intermediateThrowEvent>
```

### PR DESCRIPTION
**Pull request by lzgabel**
## Description

* add the new Zeebe extension element `publishMessage` with the attributes
    * `messageId`
    *  `correlationKey`
    *  `timeToLive`
* extend the builder of send tasks and throw events to define the publish message

## Related issues
closes #12878

## Definition of Done

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by lzgabel** [comment]
Hi @korthout. Please take a moment to review this feature. 🙇

**Review by korthout — CHANGES_REQUESTED** [review]
We're currently defining what is needed for this topic in:
- #12878 

@lzgabel Since I receive a notification about this pull request every moring, I'm now requesting the changes explicitly. This should stop the notification from pinging me every day. Sorry for the noise.

**Comment by lzgabel** [comment]
> Since I receive a notification about this pull request every moring, I'm now requesting the changes explicitly. This should stop the notification from pinging me every day. 

@korthout Sorry about that. :bow:
I'll convert it to a draft and adjust it next week.

**Comment by lzgabel** [comment]
Hi @korthout. Please take a moment to review this feature when you have time. 🙇

**Comment by korthout** [comment]
Thanks @lzgabel. I'll do my best to review it ASAP, but I have some other priorities. I'll find time for it in the next few days. And as always, thanks for the awesome contribution ❤️

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/MessageValidator.java [hunk 1]
```diff
@@ -112,6 +117,26 @@ private boolean isReferredByCatchEvent(final Message element) {
...
+        getAllElementsByType(element, IntermediateThrowEvent.class);
+
+    final Collection<EndEvent> endEvents = getAllElementsByType(element, EndEvent.class);
+
+    return Stream.concat(intermediateCatchEvents.stream(), endEvents.stream())
+        .flatMap(i -> i.getEventDefinitions().stream())
+        .anyMatch(
+            e ->
+                e instanceof MessageEventDefinition
+                    && ((MessageEventDefinition) e).getMessage() == element);
```
Comment (korthout) [inline bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/validation/zeebe/MessageValidator.java:131]: 💭 Not sure if this is actually better, but here's an alternative:

```java
.filter(MessageEventDefinition.class::isInstance)
.map(MessageEventDefinition.class::cast)
.anyMatch(e -> e.getMessage() == element);
```

And another 🤔 Maybe I like this one better

```java
.filter(MessageEventDefinition.class::isInstance)
.anyMatch(e -> ((MessageEventDefinition) e).getMessage() == element);
```

Feel free to ignore this

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @lzgabel 👏 

It looks great! I just have one request.

Please have a look 🙇

**Comment by lzgabel** [comment]
Hi @korthout. Please check again.  I'll squash all commits when you're done checking this out. 🙇

**Review by korthout — APPROVED** [review]
Thanks @lzgabel ❤️ 

LGTM 👍

**Comment by lzgabel** [comment]
Hi @korthout. I've squashed all commits. Please check this out. :heart:

**Review by korthout — APPROVED** [review]
Thanks again @lzgabel ✉️ 

LGTM 👍 

bors merge

**Pull request merged**